### ISSUE TITLE
Rustc isn’t warning about usage of and even suggests importing #[doc(hidden)] items.

### ISSUE THREAD

**Issue by steffahn** [rust-lang/rust#80948]
Motivated by [this thread](https://users.rust-lang.org/t/serde-compatibility-broken/54041) on URLO, I’m wondering if we shouldn’t add warnings for using items that are hidden in documentations, in particular hidden modules (but e.g. also hidden fields that can be used to make structs with public fields extensible while still supporting `Struct { foo: 1, bar: 2 ..Struct::default() }`-style initialization).

Even more unfortunate is how rustc even _suggests_ hidden re-export modules, e.g. in [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2b8190cd176daec757c7e8eeff4db781), the compiler suggests, among other things, to add `use pin_utils::core_reexport::fmt::Formatter;` or `use serde::export::Formatter;`, which are semver-hazards.
```
   Compiling playground v0.0.1 (/playground)
error[E0412]: cannot find type `Formatter` in this scope
 --> src/lib.rs:2:16
  |
2 | pub fn foo(_f: Formatter) {}
  |                ^^^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
2 | use core::fmt::Formatter;
  |
2 | use env_logger::fmt::Formatter;
  |
2 | use pin_utils::core_reexport::fmt::Formatter;
  |
2 | use serde::export::Formatter;
  |
    and 3 other candidates
```

@rustbot modify labels: C-enhancement, T-compiler, A-diagnostics, A-suggestion-diagnostics, D-invalid-suggestion

_(regarging `D-invalid-suggestion`: I **would** consider code that explicitly uses unstable internal (for macros) re-exports of another crate to be “incorrect”)_

**Comment by jyn514** [comment]
Relevant code: https://github.com/rust-lang/rust/blob/fc9944fe84a683f0450c0921a935456e51b1c3ae/compiler/rustc_resolve/src/diagnostics.rs#L1651

**Comment by kpreid** [comment]
I like the idea of having warnings for hidden items — I work on a set of crates with hidden “I don't think this is good public API” items, and if there was a lint for me to explicitly `allow`, then it would help me make sure I'm not accidentally making a library where _necessary_ items are hidden. The allow would act sort of as an “unsafe-for-API-usability block” which I can think carefully about and audit just like I can audit `unsafe {}` blocks.

It's kind-of promoting `doc(hidden)` to be an “official” part of the visibility system which it already is unofficially.
